### PR TITLE
INGK-987 Allow users to subscribe to register updates

### PR DESCRIPTION
**Pull request by MartinALL91**
### Description

Add methods to subscribe/unsubscribe to register value updates.

Fixes # INGK-987

### Type of change

- Add the register_update_subscribe method.
- Add the register_update_unsubscribe method.
- Remove unnecessary checks in the convert_dtype_to_bytes method.

### Tests
- [x] Add new unit tests if it applies.
- [x] Run tests.

### Documentation

Please update the documentation.

- [x] Update docstrings of every function, method or class that change.
- [x] Add the changes at the `[Unreleased]` section of the [CHANGELOG](CHANGELOG.md).

### Code formatting and linting

- [x] Use the ruff package to format the code: `ruff format ingenialink tests`.
- [x] Use the ruff package to lint the code: `ruff check ingenialink`.

### Others

- [x] Set fix version field in the Jira issue.

### CODE REVIEW

File: ingenialink/utils/_utils.py [hunk 1]
```diff
@@ -236,10 +236,6 @@ def convert_dtype_to_bytes(data: Union[int, float, str, bytes], dtype: REG_DTYPE
         and not isinstance(data, bytes)
     ):
         raise ValueError(f"Invalid value. Expected values: {VALID_BIT_REGISTER_VALUES}, got {data}")
-    if dtype == REG_DTYPE.BYTE_ARRAY_512:
```
Comment (polfeliu) [inline ingenialink/utils/_utils.py:239]: Is this change needed?

File: ingenialink/servo.py [hunk 2]
```diff
@@ -1250,11 +1251,14 @@ def write(
...
-        if not isinstance(data, bytes):
-            value = convert_dtype_to_bytes(data, _reg.dtype)
+        if isinstance(data, bytes):
+            data_bytes = data
         else:
-            value = data
-        self._write_raw(_reg, value, **kwargs)
+            data_bytes = convert_dtype_to_bytes(data, _reg.dtype)
+        self._write_raw(_reg, data_bytes, **kwargs)
+        if isinstance(data, bytes):
```
[the remark this answers is not in the document]
Comment (MartinALL91) [reply ingenialink/servo.py:1259]: The only case where the data is passed directly as bytes is when the disturbance data is written. I don't think we need to store this in the register models.

File: ingenialink/servo.py [hunk 3]
```diff
@@ -1283,8 +1287,43 @@ def read(
...
+
+        Args:
+            reg: Updated register.
+            data: Updated value.
+
+        """
+        for callback in self.__register_update_observers:
+            callback(
+                self,
+                Register(reg.dtype, reg.access, reg.identifier, subnode=reg.subnode, storage=data),
```
[the remark this answers is not in the document]
Comment (MartinALL91) [reply ingenialink/servo.py:1324]: The register object can hold multiple information (UID, subnode, value) in a single object.

File: ingenialink/utils/_utils.py (hunk 1 above)
Comment (MartinALL91) [reply ingenialink/utils/_utils.py:239]: Not really but it does not make sense to have it.

File: ingenialink/servo.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (GullinBustin) [reply ingenialink/servo.py:1324]: I don't understand, creating a new register we lose info, such as register type, enums, etc... 
In addition, should we also send it if the notification comes from a read or a write?

File: ingenialink/utils/_utils.py (hunk 1 above)
Comment (GullinBustin) [reply ingenialink/utils/_utils.py:239]: because it is already bytes? but maybe in some case, you don't know the register type but you want to convert data to bytes, in this case now it will raise an exception

File: ingenialink/servo.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (MartinALL91) [reply ingenialink/servo.py:1324]: I did it this way to avoid modifying the servo's register objects. But now I think there should be no problems since all registers will be reread before storing a configuration file.

**Review by GullinBustin — APPROVED** [review]

File: ingenialink/servo.py [hunk 4]
```diff
@@ -1284,8 +1286,45 @@ def read(
...
+        """Notify a register update to the observers.
+        The updated value is stored in the register's storage attribute.
+
+        Args:
+            reg: Updated register.
+            data: Updated value.
+
+        """
+        reg.storage_valid = True
+        reg.storage = data
```
Comment (GullinBustin) [inline ingenialink/servo.py:1321]: Talking with @polfeliu , we thought that the storage attribute in the register should not be used. Dictionary and Register classes should be static. For the moment we don't need a cache for the registers values, and if we need it in the future maybe the registers are not the best class to implement it.

**Review by GullinBustin — CHANGES_REQUESTED** [review]

File: ingenialink/servo.py (hunk 4 above)
Comment (MartinALL91) [reply ingenialink/servo.py:1321]: Done

File: tests/test_servo.py [hunk 5]
```diff
@@ -603,3 +616,30 @@ def test_disturbance_overflow(connect_to_slave, pytestconfig):
...
+    servo.register_update_subscribe(register_update_callback.register_update_test)
+
+    previous_reg_value = servo.read(user_over_voltage_uid, subnode=1)
+    assert register_update_callback.call_count == 1
+    assert register_update_callback.servo == servo
+    assert register_update_callback.register.identifier == user_over_voltage_uid
+    assert register_update_callback.value == previous_reg_value
+
+    new_reg_value = 100
+    servo.write(user_over_voltage_uid, data=new_reg_value, subnode=1)
```
Comment (polfeliu) [inline tests/test_servo.py:637]: if you want you could also test that when a write fails no notification is sent.
reading the code I think is properly implemented

**Review by GullinBustin — APPROVED** [review]
